okhttp/src/main/kotlin/okhttp3/internal/tls/DnsUtils.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package okhttp3.internal.tls
+
+/**
+ * A collection of utilities relating to Domain Name System.
+ *
+ * https://raw.githubusercontent.com/apache/httpcomponents-client/master/httpclient5/src/main/java/org/apache/hc/client5/http/utils/DnsUtils.java
+ */
+internal object DnsUtils {
+  private fun isIA5(c: Char): Boolean {
+    return c < 128.toChar()
+  }
+
+  private fun isUpper(c: Char): Boolean {
+    return c in 'A'..'Z'
+  }
+
+  // https://tools.ietf.org/html/rfc2459#section-4.2.1.11
+  fun normalizeIA5String(s: String): String {
+    var pos = 0
+    var remaining = s.length
+    while (remaining > 0) {
+      // TODO enable if we agree to enforce here
+//      check (isIA5(s[pos])) {
```

Comment:
@swankjesse if we trust our overall logic, we should enforce this. Tests like verifyMultipleCn will fail and need to be adjusted.

Reply:
This is an addition to the original file which was converted from java.

Reply:
I don’t think we need this? I also don’t think Apache HTTP components needs it? I think a better solution is to simply require the input argument is already lowercase ASCII?

Reply:
In particular, `verifyHostname()` is not a public API. We only need this defense if we’re passing in strings ourselves that need this transformation.

Reply:
Yep, as far as I can tell 100% of the hosts we pass to `verify()` have already passed through `okhttp3.internal.HostnamesKt#toCanonicalHost` which is ASCII-only.

We could potentially document or enforce that constraint?

Reply:
> We could  ... enforce that constraint?

That is this block.  I'll follow up on later PR.

Reply:
verify() is public API though.  While we package OkHostnameVerifier in an internal package, it's a public API exposed very obviously in OkHttpClient via the hostnameVerifier() accessor.